### ISSUE TITLE
Prevent local imports from hiding imported symbols

### ISSUE THREAD

**Issue by dlangBugzillaToGithub** [dlang/dmd#19272]
### CC List

* Bolpat

### Description

````markdown
As issue 10378 has been closed even though problems remain, I'm opening a new issue.

The following code demonstrates a hijacking problem in the current local import lookup rules:

import std.stdio;

string readAndLog(string filename){
    import std.file;
    auto text=readText(filename);
    write(filename," read successfully!
");
    return text;
}

void main(){
    writeln(readAndLog("important_data.txt"));
}

Upon execution, this code will delete the important data.

Local imports should overload against less deeply nested imports. In this example, std.stdio.write and std.file.write should be treated as part of the same overload set within readAndLog. Otherwise, there is a hijacking risk.
````

**Comment by dlangBugzillaToGithub** [comment]
#### qs.il.paperinik commented on 2020-11-22T03:31:13Z



````markdown
(In reply to timon.gehr from comment #0)
> Local imports should overload against less deeply nested imports. In this
> example, std.stdio.write and std.file.write should be treated as part of the
> same overload set within readAndLog. Otherwise, there is a hijacking risk.

You imported two `write` functions (overloaded sets) that weren't intended to be used next to each other. Call compiles in both sets, so by the spec [1], it's an ambiguous call. The spec does not say that "more local" imports shadow "less local" imports for functions. While [2] says that imports can shadow local definitions, [1] clearly says how overloaded sets are to be handled. At best, [1] is correct and at worst, the spec contradicts itself.

[1] https://dlang.org/spec/function.html#overload-sets
[2] https://dlang.org/spec/module.html#scoped_imports

TL;DR: It should be an ambiguity error, nothing else.
````

**Comment by dlangBugzillaToGithub** [comment]
#### timon.gehr commented on 2020-11-22T03:51:19Z



````markdown
(In reply to Bolpat from comment #1)
> (In reply to timon.gehr from comment #0)
> > Local imports should overload against less deeply nested imports. In this
> > example, std.stdio.write and std.file.write should be treated as part of the
> > same overload set within readAndLog. Otherwise, there is a hijacking risk.
> 
> ...
> 
> TL;DR: It should be an ambiguity error, nothing else.

Yes absolutely. Reading it now, the original report was not phrased very well. std.stdio.write and std.file.write should remain in separate overload sets, but they should overload against each other so that there is an ambiguity error in this case.
````

**Comment by WalterBright** [comment]
I presume the idea then is all imports in nested scopes should overload against the global imports? That it's not possible for an import to override an import in an enclosing scope?

**Comment by Bolpat** [comment]
> I presume the idea then is all imports in nested scopes should overload against the global imports? That it's not possible for an import to override an import in an enclosing scope?

I guess so. Unless using `import m : s`, which brings direct attention to `s`, it shouldn’t matter “how locally” you imported something. I’m not sure what the rules are exactly, but maybe we can sell `import m : s` as `alias s = imported!"m".s`, i.e. something local.
